src/R2N.jl

```diff
@@ -0,0 +1,387 @@
+  - `stats.elapsed_time`: elapsed time in seconds.
+
+# Examples
+```jldoctest
+using JSOSolvers, ADNLPModels
+nlp = ADNLPModel(x -> sum(x.^2), ones(3))
+stats = R2N(nlp)
+
+# output
+
+"Execution stats: first-order stationary"
+```
+
+```jldoctest
+using JSOSolvers, ADNLPModels
+nlp = ADNLPModel(x -> sum(x.^2), ones(3))
+solver = R2NSolver(nlp);
+stats = solve!(solver, nlp)
+
+# output
+
+"Execution stats: first-order stationary"
+```
+"""
+mutable struct R2NSolver{
+  T,
+  V,
+  Op <: AbstractLinearOperator{T},
+  Op2 <: AbstractLinearOperator{T},
+  Sub <: Union{KrylovSolver{T, T, V}, ShiftedLBFGSSolver},
+} <: AbstractOptimizationSolver
+  x::V
+  cx::V
+  gx::V
+  gn::V
+  σ::T
+  μ::T
+  H::Op
+  shifted_H::Op2
+  Hs::V
+  s::V
+  obj_vec::V # used for non-monotone behaviour
+  subsolver_type::Sub
+  cgtol::T
+end
+
+function R2NSolver(
+  nlp::AbstractNLPModel{T, V};
+  non_mono_size = 1,
+  subsolver_type::Union{Type{<:KrylovSolver}, Type{ShiftedLBFGSSolver}} = MinresSolver,
+) where {T, V}
+  nvar = nlp.meta.nvar
+  x = V(undef, nvar) # vs similar(nlp.meta.x0)
+  cx = V(undef, nvar)
+  gx = V(undef, nvar)
+  gn = isa(nlp, QuasiNewtonModel) ? V(undef, nvar) : V(undef, 0)
+  Hs = V(undef, nvar)
+  H = isa(nlp, QuasiNewtonModel) ? nlp.op : hess_op!(nlp, x, Hs)
+  shifted_H = LinearOperator(Matrix{T}(I, nvar, nvar))
+  Op = typeof(H)
+  Op2 = typeof(shifted_H)
+  σ = zero(T) # init it to zero for now 
+  μ = zero(T) # init it to zero for now
+  s = V(undef, nvar)
+  cgtol = one(T) # must be ≤ 1.0
+  obj_vec = fill(typemin(T), non_mono_size)
+  subsolver =
+    isa(subsolver_type, Type{ShiftedLBFGSSolver}) ? subsolver_type() : subsolver_type(nvar, nvar, V)
+
+  Sub = typeof(subsolver)
+  return R2NSolver{T, V, Op, Op2, Sub}(
+    x,
+    cx,
+    gx,
+    gn,
+    σ,
+    μ,
+    H,
+    shifted_H,
+    Hs,
+    s,
+    obj_vec,
+    subsolver,
+    cgtol,
+  )
+end
+
+function SolverCore.reset!(solver::R2NSolver{T}) where {T}
+  fill!(solver.obj_vec, typemin(T))
+  reset!(solver.H)
+  solver
+end
+function SolverCore.reset!(solver::R2NSolver{T}, nlp::AbstractNLPModel) where {T}
+  fill!(solver.obj_vec, typemin(T))
+  @assert (length(solver.gn) == 0) || isa(nlp, QuasiNewtonModel)
+  solver.H = isa(nlp, QuasiNewtonModel) ? nlp.op : hess_op!(nlp, solver.x, solver.Hs)
+  solver
+end
+
+@doc (@doc R2NSolver) function R2N(
+  nlp::AbstractNLPModel{T, V};
+  subsolver_type::Union{Type{<:KrylovSolver}, Type{ShiftedLBFGSSolver}} = MinresSolver,
+  non_mono_size = 1,
+  kwargs...,
+) where {T, V}
+  solver = R2NSolver(nlp; non_mono_size = non_mono_size, subsolver_type = subsolver_type)
+  return solve!(solver, nlp; non_mono_size = non_mono_size, kwargs...)
+end
+
+function SolverCore.solve!(
+  solver::R2NSolver{T, V},
+  nlp::AbstractNLPModel{T, V},
+  stats::GenericExecutionStats{T, V};
+  callback = (args...) -> nothing,
+  x::V = nlp.meta.x0,
+  atol::T = √eps(T),
+  rtol::T = √eps(T),
+  η1 = T(0.0001),
+  η2 = T(0.001),
+  λ = T(2), #>1
```

Comment:
```suggestion
  λ = T(2),
```
add this in the Keyword arguments in the docstring